Add animationImages support when using SDAnimatedImageView

MoZhouqi opened the pull request:
### New Pull Request Checklist

* [x] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [x] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [x] I have searched for a similar pull request in the [project](https://github.com/rs/SDWebImage/pulls) and found none

* [x] I have updated this branch with the latest master to avoid conflicts (via merge from master or rebase)
* [x] I have added the required tests to prove the fix/feature I am adding
* [x] I have updated the documentation (if necessary)
* [x] I have run the tests and they pass
* [x] I have run the lint and it passes (`pod lib lint`)

This merge request fixes / refers to the following issues: ...

### Pull Request Description
Add `animationImages` support when using `SDAnimatedImageView`.
...

kinarobin commented on SDWebImage/Core/SDAnimatedImageView.m:
```diff
@@ -194,6 +194,13 @@ - (void)setImage:(UIImage *)image
     }
 }
 
+- (void)setAnimationImages:(NSArray<UIImage *> *)animationImages
```
Maybe this method should apply to `#if SD_UIKIT`.

kinarobin commented on SDWebImage/Core/SDAnimatedImageView.m:
```diff
@@ -421,7 +428,7 @@ - (void)updateShouldAnimate
 #endif
-    self.shouldAnimate = self.player && isVisible;
+    self.shouldAnimate = (self.player || self.animationImages) && isVisible;
 }
```
Same! `self.animationImages` Just in UIKit, you should think about AppKit.

MoZhouqi commented:
@kinarobin Thanks for your suggestion, I have updated the code.

dreampiggy commented:
I wonder, what cause this issues ?
We override the super startAnimating: and setImage:, and setAnimations also use the super implementation.

If you use `UIImageView.aniamtionImages`, it shoud work as normal, because it's super's implementation

MoZhouqi commented:
> I wonder, what cause this issue ?
> 
> We override the super startAnimating: and setImage:, and setAnimations also use the super implementation.
> 
> If you use `UIImageView.aniamtionImages`, it shoud work as normal, because it's super's implementation

For example:
```objc
SDAnimatedImageView *imageView = [SDAnimatedImageView alloc] initWithFrame:frame];
imageView.animationImages = animationImages;
[imageView startAnimating];
imageView.hidden = NO;
```
There are two issues with the above code:
1. Although the default value of `autoPlayAnimatedImage` is `YES`, the animation is not playing automatically after setting a non-nil value on the `animationImages` property.
2. When `setHidden:` method is called,  `checkPlay` method will be invoked and then the animation will be stopped.

dreampiggy commented:
> > I wonder, what cause this issue ?
> > We override the super startAnimating: and setImage:, and setAnimations also use the super implementation.
> > If you use `UIImageView.aniamtionImages`, it shoud work as normal, because it's super's implementation
> 
> For example:
> 
> ```objc
> SDAnimatedImageView *imageView = [SDAnimatedImageView alloc] initWithFrame:frame];
> imageView.animationImages = animationImages;
> [imageView startAnimating];
> imageView.hidden = NO;
> ```
> 
> There are two issues with the above code:
> 
> 1. Although the default value of `autoPlayAnimatedImage` is `YES`, the animation is not playing automatically after setting a non-nil value on the `animationImages` property.
> 2. When `setHidden:` method is called,  `checkPlay` method will be invoked and then the animation will be stopped.

I see... This can be updated into the Unit Test (SDWebImage Tetsts), so that can ensure this will not be broken in later version.

MoZhouqi commented:
> > > I wonder, what cause this issue ?
> > > We override the super startAnimating: and setImage:, and setAnimations also use the super implementation.
> > > If you use `UIImageView.aniamtionImages`, it shoud work as normal, because it's super's implementation
> > 
> > 
> > For example:
> > ```objc
> > SDAnimatedImageView *imageView = [SDAnimatedImageView alloc] initWithFrame:frame];
> > imageView.animationImages = animationImages;
> > [imageView startAnimating];
> > imageView.hidden = NO;
> > ```
> > 
> > 
> > There are two issues with the above code:
> > 
> > 1. Although the default value of `autoPlayAnimatedImage` is `YES`, the animation is not playing automatically after setting a non-nil value on the `animationImages` property.
> > 2. When `setHidden:` method is called,  `checkPlay` method will be invoked and then the animation will be stopped.
> 
> I see... This can be updated into the Unit Test (SDWebImage Tetsts), so that can ensure this will not be broken in later version.

I've added the test cases for `animationImages`.

dreampiggy reviewed (changes requested):
I have another way to solve this problem, by leaving all the control to UIKit, we don't touch.

Remove all the changes to `SDAnimatedImageView`, only modify this method:

```objective-c
/// Check if it should be played
- (void)checkPlay
{
    // Only handle for SDAnimatedImage, leave UIAnimatedImage or animatedImages for super implementation control
    if (self.player && self.autoPlayAnimatedImage) {
        [self updateShouldAnimate];
        if (self.shouldAnimate) {
            [self startAnimating];
        } else {
            [self stopAnimating];
        }
    }
}
```

I have tested this also match your behavior, but it's more suitable and maintainable. I think, **SDAnimatedImageView should be a drop-in replacement for UIImageView**. So if super class implementation do something, we should call their instead and only handle the case we added (That `SDAnimatedImagePlayer` or `SDAnimatedImage`), not assume current iOS's behavior.

See:

![image](https://user-images.githubusercontent.com/6919743/97655327-c8727a80-1a9f-11eb-8e4d-ec14e2d919b3.png)

MoZhouqi commented:
> I have another way to solve this problem, by leaving all the control to UIKit, we don't touch.
> 
> Remove all the changes to `SDAnimatedImageView`, only modify this method:
> 
> ```objc
> /// Check if it should be played
> - (void)checkPlay
> {
>     // Only handle for SDAnimatedImage, leave UIAnimatedImage or animatedImages for super implementation control
>     if (self.player && self.autoPlayAnimatedImage) {
>         [self updateShouldAnimate];
>         if (self.shouldAnimate) {
>             [self startAnimating];
>         } else {
>             [self stopAnimating];
>         }
>     }
> }
> ```
> 
> I have tested this also match your behavior, but it's more suitable and maintainable. I think, **SDAnimatedImageView should be a drop-in replacement for UIImageView**. So if super class implementation do something, we should call their instead and only handle the case we added (That `SDAnimatedImagePlayer` or `SDAnimatedImage`), not assume current iOS's behavior.
> 
> For this issue, I think this `autoPlayAnimatedImage` should only effect `SDAnimatedImage`, we can update the code comment in header files about this.
> 
> See:
> 
> ![image](https://user-images.githubusercontent.com/6919743/97655327-c8727a80-1a9f-11eb-8e4d-ec14e2d919b3.png)

Yes I see, it is better, I have updated the code, the test cases and the code comment in header files.

MoZhouqi commented:
@dreampiggy Yes I see, it is better, I have updated the code, the test cases and the code comment in header files.

dreampiggy reviewed (approved):
Now it's better :)

MoZhouqi commented:
> Now it's better :)

🍻 Yes, the previous treatment is not perfect, I should also deal with other properties properly, such as `currentFrame`.
